Flaky TestWaypointAsEgressGateway Integration Test

### Is this the right place to submit this?

- [x] This is not a security vulnerability or a crashing bug
- [x] This is not a question about how to use Istio

### Bug Description

We are observing flakiness related to TestWaypointAsEgressGateway on [release-1.28](https://prow.istio.io/view/gs/istio-prow/logs/integ-ambient-owned-cni_istio_release-1.28_postsubmit/1978846698629238784) and [master](https://prow.istio.io/view/gs/istio-prow/logs/integ-k8s-126_istio_postsubmit/1978658451374477312)

```bash
2025-10-16T03:18:13.681782Z	info	tf	=== BEGIN: Test: 'ambient[TestWaypointAsEgressGateway/resolution_none_10.96.216.39/from_workload-addressed-waypoint]' ===
    waypoint_test.go:524: 2 errors occurred:
        	* failed calling workload-addressed-waypoint (cluster=cluster-0)->'http://10.96.216.39:8080': call failed from workload-addressed-waypoint (cluster=cluster-0) to http://10.96.216.39:8080 (using http): response[0]: expected hostname external-v2-7c4bcdbbf9-kj7sg, received external-v1-bcdff56bc-2brzl
        	* failed calling workload-addressed-waypoint (cluster=cluster-0)->'http://10.96.216.39:8080': call failed from workload-addressed-waypoint (cluster=cluster-0) to http://10.96.216.39:8080 (using http): response[0]: expected hostname external-v2-7c4bcdbbf9-kj7sg, received external-v1-bcdff56bc-2brzl
```

```bash
FAIL: TestWaypointAsEgressGateway/resolution_none_10.96.90.117 (235.68s)
        --- FAIL: TestWaypointAsEgressGateway/resolution_none_10.96.90.117/from_workload-addressed-waypoint (77.22s)
        --- FAIL: TestWaypointAsEgressGateway/resolution_none_10.96.90.117/from_service-addressed-waypoint (73.27s)
        --- FAIL: TestWaypointAsEgressGateway/resolution_none_10.96.90.117/from_captured (72.38s)
```

### Version

```Text
master
1.28 branch
```

### Additional Information

_No response_

--- a/tests/integration/ambient/waypoint_test.go
+++ b/tests/integration/ambient/waypoint_test.go
@@ -27,6 +27,7 @@ import (
 
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/types"
 	k8s "sigs.k8s.io/gateway-api/apis/v1"
 
@@ -464,6 +465,7 @@ type externalSubsetService struct {
 }
 
 // servicesForSubsets is a helper function to create a kubernetes service for all subsets of a service
+// Assumes a single pod per subset
 func servicesForSubsets(t framework.TestContext, instance echo.Instance) []externalSubsetService {
 	ns := instance.Config().Namespace.Name()
 	services := []externalSubsetService{}
@@ -495,13 +497,27 @@ func servicesForSubsets(t framework.TestContext, instance echo.Instance) []exter
 		if err != nil {
 			t.Fatalf("failed to create service %s: %v", svc.Name, err)
 		}
-		t.Cleanup(func() {
+
+		t.CleanupConditionally(func() {
 			t.Clusters().Default().Kube().CoreV1().Services(ns).Delete(context.TODO(), s.Name, metav1.DeleteOptions{})
 		})
+
+		pods, err := t.Clusters().Default().Kube().CoreV1().Pods(ns).List(context.TODO(), metav1.ListOptions{
+			LabelSelector: labels.SelectorFromSet(map[string]string{
+				"app":     instance.Config().Service,
+				"version": subset.Version,
+			}).String(),
+		})
+		if err != nil {
+			t.Fatalf("failed to list pods: %v", err)
+		}
+		if len(pods.Items) != 1 {
+			t.Fatalf("expected 1 pod found for service %s, got %d", svc.Name, len(pods.Items))
+		}
 		services = append(services, externalSubsetService{
 			Name:      s.Name,
 			SubsetKey: subset.Version,
-			Hostname:  instance.WorkloadsOrFail(t)[0].PodName(),
+			Hostname:  pods.Items[0].Name,
 			ClusterIP: s.Spec.ClusterIP,
 		})
 	}
@@ -569,7 +585,6 @@ spec:
 
 			// Use the first subset service for testing consistency
 			subsetService := servicesForSubsets(t, external)[0]
-
 			resolutionNoneServiceEntry := `apiVersion: networking.istio.io/v1
 kind: ServiceEntry
 metadata:

